### PR TITLE
[SLP]Add subvector vectorization for non-load nodes

### PR DESCRIPTION
**Pull request by alexey-bataev**
Previously SLP vectorize supported clustered vectorization for loads
only. This patch adds support for "clustered" vectorization for other
instructions.
If the buildvector node contains "clusters", which can be vectorized
separately and then inserted into the resulting buildvector result, it
is better to do, since it may reduce the cost of the vector graph and
produce better vector code.
The patch does some analysis, if it is profitable to try to do this kind
of extra vectorization. It checks the scalar instructions and its
operands and tries to vectorize them only if they result in a better
graph.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -8447,47 +8454,147 @@ getGEPCosts(const TargetTransformInfo &TTI, ArrayRef<Value *> Ptrs,
...
+    SmallVector<SmallVector<std::pair<Value *, Value *>>> Candidates;
+    for (unsigned Op : seq<unsigned>(S.MainOp->getNumOperands()))
+      Candidates.emplace_back().emplace_back(I1->getOperand(Op),
+                                             I2->getOperand(Op));
+    return all_of(Candidates, [this](
+                                  ArrayRef<std::pair<Value *, Value *>> Cand) {
+      return all_of(Cand,
+                    [](const std::pair<Value *, Value *> &P) {
+                      return isa<Constant>(P.first) ||
+                             isa<Constant>(P.second) || (P.first == P.second);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8476]: Should this be `(isa<Constant>(P.first) && isa<Constant>(P.second)) || (P.first == P.second)` ?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8476]: No, the logic is correct here

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8476]: I will update the comment and remove extra parens

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by aeubanks** [comment]
I believe we're seeing a miscompile with this change, still reducing

**Comment by aeubanks** [comment]
here's a diff of the relevant function after slp-vectorizer with and without this patch reverted, and with inst names all set to `%0` for minimal diffing purposes. I'll keep digging, but perhaps you can see something from this diff?
```
177a178,181
>   %0 = insertelement <2 x i32> poison, i32 %lshr166, i32 0
>   %0 = insertelement <2 x i32> %0, i32 %trunc176, i32 1
>   %0 = and <2 x i32> %0, <i32 255, i32 255>
>   %0 = icmp eq <2 x i32> %0, zeroinitializer
188,194c192
<   %0 = insertelement <2 x i32> poison, i32 %lshr166, i32 0
<   %0 = insertelement <2 x i32> %0, i32 %trunc176, i32 1
<   %0 = and <2 x i32> %0, <i32 255, i32 255>
<   %0 = icmp eq <2 x i32> %0, zeroinitializer
<   %0 = zext <2 x i8> %0 to <2 x i32>
<   %0 = shufflevector <2 x i32> %0, <2 x i32> %0, <8 x i32> <i32 poison, i32 poison, i32 0, i32 1, i32 poison, i32 poison, i32 2, i32 3>
<   %0 = insertelement <8 x i32> %0, i32 %trunc111, i32 0
---
>   %0 = insertelement <8 x i32> poison, i32 %trunc111, i32 0
197a196,199
>   %0 = shufflevector <2 x i32> %0, <2 x i32> poison, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
>   %0 = shufflevector <8 x i32> %0, <8 x i32> %0, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 4, i32 5, i32 8, i32 9>
>   %0 = sext <2 x i8> %0 to <2 x i32>
>   %0 = call <8 x i32> @llvm.vector.insert.v8i32.v2i32(<8 x i32> %0, <2 x i32> %0, i64 2)
204a207,208
>   %0 = shufflevector <2 x i1> %0, <2 x i1> poison, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
>   %0 = shufflevector <8 x i1> %0, <8 x i1> %0, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 4, i32 5, i32 8, i32 9>
206d209
<   %0 = call <8 x i1> @llvm.vector.insert.v8i1.v2i1(<8 x i1> %0, <2 x i1> %0, i64 6)
238c241,243
<   %0 = call <8 x i64> @llvm.vector.insert.v8i64.v4i64(<8 x i64> poison, <4 x i64> %0, i64 0)
---
>   %0 = shufflevector <2 x i64> %0, <2 x i64> poison, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
>   %0 = shufflevector <8 x i64> %0, <8 x i64> poison, <8 x i32> <i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 0, i32 1>
>   %0 = call <8 x i64> @llvm.vector.insert.v8i64.v4i64(<8 x i64> %0, <4 x i64> %0, i64 0)
240d244
<   %0 = call <8 x i64> @llvm.vector.insert.v8i64.v2i64(<8 x i64> %0, <2 x i64> %0, i64 6)
```

**Comment by alexey-bataev** [comment]
> here's a diff of the relevant function after slp-vectorizer with and without this patch reverted, and with inst names all set to `%0` for minimal diffing purposes. I'll keep digging, but perhaps you can see something from this diff?
> 
> ```
> 177a178,181
> >   %0 = insertelement <2 x i32> poison, i32 %lshr166, i32 0
> >   %0 = insertelement <2 x i32> %0, i32 %trunc176, i32 1
> >   %0 = and <2 x i32> %0, <i32 255, i32 255>
> >   %0 = icmp eq <2 x i32> %0, zeroinitializer
> 188,194c192
> <   %0 = insertelement <2 x i32> poison, i32 %lshr166, i32 0
> <   %0 = insertelement <2 x i32> %0, i32 %trunc176, i32 1
> <   %0 = and <2 x i32> %0, <i32 255, i32 255>
> <   %0 = icmp eq <2 x i32> %0, zeroinitializer
> <   %0 = zext <2 x i8> %0 to <2 x i32>
> <   %0 = shufflevector <2 x i32> %0, <2 x i32> %0, <8 x i32> <i32 poison, i32 poison, i32 0, i32 1, i32 poison, i32 poison, i32 2, i32 3>
> <   %0 = insertelement <8 x i32> %0, i32 %trunc111, i32 0
> ---
> >   %0 = insertelement <8 x i32> poison, i32 %trunc111, i32 0
> 197a196,199
> >   %0 = shufflevector <2 x i32> %0, <2 x i32> poison, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
> >   %0 = shufflevector <8 x i32> %0, <8 x i32> %0, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 4, i32 5, i32 8, i32 9>
> >   %0 = sext <2 x i8> %0 to <2 x i32>
> >   %0 = call <8 x i32> @llvm.vector.insert.v8i32.v2i32(<8 x i32> %0, <2 x i32> %0, i64 2)
> 204a207,208
> >   %0 = shufflevector <2 x i1> %0, <2 x i1> poison, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
> >   %0 = shufflevector <8 x i1> %0, <8 x i1> %0, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 4, i32 5, i32 8, i32 9>
> 206d209
> <   %0 = call <8 x i1> @llvm.vector.insert.v8i1.v2i1(<8 x i1> %0, <2 x i1> %0, i64 6)
> 238c241,243
> <   %0 = call <8 x i64> @llvm.vector.insert.v8i64.v4i64(<8 x i64> poison, <4 x i64> %0, i64 0)
> ---
> >   %0 = shufflevector <2 x i64> %0, <2 x i64> poison, <8 x i32> <i32 0, i32 1, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
> >   %0 = shufflevector <8 x i64> %0, <8 x i64> poison, <8 x i32> <i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 0, i32 1>
> >   %0 = call <8 x i64> @llvm.vector.insert.v8i64.v4i64(<8 x i64> %0, <4 x i64> %0, i64 0)
> 240d244
> <   %0 = call <8 x i64> @llvm.vector.insert.v8i64.v2i64(<8 x i64> %0, <2 x i64> %0, i64 6)
> ```

Hmm, hard to tell, but the only significant difference I see is zext/sext

**Comment by alexfh** [comment]
Here is the test that fails after this patch: https://github.com/google/highway/blob/master/hwy/tests/div_test.cc

The failure looks like this:
```
[ RUN      ] HwyDivTestGroup/HwyDivTest.TestAllIntegerDiv/EMU128


u8x8 expect [2+ ->]:
  0x00,0x01,0x01,0x00,0x00,0x00,
u8x8 actual [2+ ->]:
  0x00,0x01,0x00,0x00,0x00,0x00,
Abort at div_test.cc:65: EMU128, u8x8 lane 4 mismatch: expected '0x01', got '0x00'.
```

**Comment by steelannelida** [comment]
I managed a godbolt reproduction: https://gcc.godbolt.org/z/oK9vhjW48

I'm not sure which compiler flags are relevant

**Comment by hiraditya** [comment]
We need more test cases for this patch.

**Comment by alexey-bataev** [comment]
> I managed a godbolt reproduction: https://gcc.godbolt.org/z/oK9vhjW48
> 
> I'm not sure which compiler flags are relevant

Ok, I almost figured out what's the cause of the miscompile. As I said, the problem is in minbitwidth analsysi and transformation, need to emit zext instead of sext. The fix expected to be soon

**Comment by aeubanks** [comment]
I tried minimizing some more by outlining code and am seeing SLPVectorizer trunc i64 to i8 then zext to i32 when the original code only truncs from i64 to i32, is that something that should happen?

**Comment by alexey-bataev** [comment]
> I tried minimizing some more by outlining code and am seeing SLPVectorizer trunc i64 to i8 then zext to i32 when the original code only truncs from i64 to i32, is that something that should happen?

Yes, minbitwidth analysis to reduce the size of the data/reg usage.

The problem must be fixed in c1b911c5798654b9b77746901919b86b3f1fe94c

**Comment by aeubanks** [comment]
> > I tried minimizing some more by outlining code and am seeing SLPVectorizer trunc i64 to i8 then zext to i32 when the original code only truncs from i64 to i32, is that something that should happen?
> 
> Yes, minbitwidth analysis to reduce the size of the data/reg usage.
> 
> The problem must be fixed in [c1b911c](https://github.com/llvm/llvm-project/commit/c1b911c5798654b9b77746901919b86b3f1fe94c)

thanks, that fixes the miscompiled test we were seeing

**Pull request merged**